### PR TITLE
Introduce `SanitizedPathBuf`

### PR DESCRIPTION
**Pull request by JunkuiZhang**
This PR introduces `sanitized_pathbuf` and `sanitized_pathbuf_string` methods to the `PathExt` trait in `crates/util/src/paths.rs`. These methods clean up paths, removing the `\\?\` on Windows.

This PR also introduces `SanitizedPathBuf`, which is simply a wrapper around `PathBuf`, but stores the `PathBuf` with the `\\?\` prefix removed.

The functions involved are as follows:
- `Path/PathBuf::sanitized_pathbuf`: On Windows, this removes the `\\?\` prefix and returns `Result<SanitizedPathBuf>`. On non-Windows platforms, this function will never return an `Err`.
- `Path/PathBuf::sanitized_pathbuf_string`: On Windows, this removes the `\\?\` prefix and returns `Result<String>`. On non-Windows platforms, it will never return an `Err`, and this function is only used on Windows for now.
- `SanitizedPathBuf::Into<PathBuf>`: This simply returns the `PathBuf` that `SanitizedPathBuf` contained.
- `PathBuf::Into<SanitizedPathBuf>`: This function is similar to `Path/PathBuf::sanitized_pathbuf`, but it does not return a `Result`, instead directly returning a `SanitizedPathBuf`.

I'm not entirely sure if this is the best solution, so any feedback is welcome. If this approach is acceptable, I'll write the relevant test cases and then this PR will be ready to merge.

cc @SomeoneToIgnore 

Release Notes:

- N/A

### CODE REVIEW

File: crates/util/src/paths.rs [hunk 1]
```diff
@@ -93,6 +97,78 @@ impl<T: AsRef<Path>> PathExt for T {
...
+            }
+            return Ok(trimmed_path_string.to_string());
+        }
+        #[cfg(not(target_os = "windows"))]
+        return Ok(path_string.to_string());
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub struct SanitizedPathBuf(PathBuf);
```
Comment (SomeoneToIgnore) [inline crates/util/src/paths.rs:142]: I've been thinking of something else, more like a state where we have both a proper path and a proper data for displaying:

```rs
struct SanitizedPathBuf {
    original: PathBuf,
    #[cfg(win)]
    trimmed: PathBuf,
}

impl std::fmt::Display for SanitizedPathBuf {
  // #[cfg(not(win))] and #[cfg(win)]
}

impl std::convert::TryFrom for SanitizedPathBuf {
  type Error = anyhow::Error;

  // #[cfg(not(win))] and #[cfg(win)]
}

// Or should we replace it with a custom `impl SlashCommandEntry`?
// Technically, we still need two paths to return, the canonicalized one and the trimmed one?
// In this API, I presumed it's the trimmed (noop trimmed on Unix) one.
impl AsRef<Path> for SlashCommandEntry {
  // #[cfg(not(win))] and #[cfg(win)]
}
```

then no Deref [semi-antipatterns](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html) are needed.

With this, we can convert things on the spot with that new constructor that might return errors.
No other ways around that constructor — there should be no way to sanitize without the error possibility.

File: crates/util/src/paths.rs [hunk 2]
```diff
@@ -224,7 +300,10 @@ impl PathWithPosition {
         let maybe_file_name_with_row_col = path.file_name().unwrap_or_default().to_string_lossy();
         if maybe_file_name_with_row_col.is_empty() {
             return Self {
-                path: Path::new(s).to_path_buf(),
+                // TODO:
+                // If `s = "text_file.txt"` some relative file path, is it okay
+                // if we trimmed the path here?
```
Comment (SomeoneToIgnore) [inline crates/util/src/paths.rs:305]: I do not understand this TODO: if the path is relative, it's not canonicalized and should not start with `\\?\`, so there's nothing to trim and it's ok?

File: crates/terminal_view/src/terminal_view.rs [hunk 3]
```diff
@@ -717,7 +717,7 @@ fn subscribe_for_terminal_events(
                             .await;
                         let paths_to_open = valid_files_to_open
                             .iter()
-                            .map(|(p, _)| p.path.clone())
+                            .map(|(p, _)| p.path.clone().into())
```
Comment (SomeoneToIgnore) [inline crates/terminal_view/src/terminal_view.rs:720]: Overall, it's an interesting idea trying making `PathWithPosition`'s path sanitized.

Yet, original issue seemed to be in the fact that `workspace.open_paths` received canonicalized paths and failed to work with them properly?
Any other "bad" input will still break things for workspace as the ctrl-click did before your fix.
This is where we should sanitize too, and for that, we have to change `fn open_path` signature.

Maybe, same idea might relate to other changes in this PR?
`trash_file`, `trash_dir`, and `file_save_dialog` seem all to accept a path, and can require a proper `SanitizedPathBuf` instead.

I find this direction interesting, as all of these require an absolute, symlink-traversed (i.e. canonicalized) path for their work, so `SanitizedPathBuf` looks natural there.
More approchable and "natural" compared to placing the `SanitizedPathBuf` into `PathWithPosition`, as that one may accept relative paths and other things that look like paths but end with :22:33 things, so not really about the same `\\?\` issue?

File: crates/util/src/paths.rs [hunk 4]
```diff
@@ -93,6 +97,78 @@ impl<T: AsRef<Path>> PathExt for T {
...
+    ///
+    /// Generally, this method should be used when the path is to be displayed to the user,
+    /// and the path should be returned from `canonicalize`.
+    fn sanitized_pathbuf_string(&self) -> anyhow::Result<String> {
+        let path_string = self.as_ref().to_string_lossy();
+        #[cfg(target_os = "windows")]
+        {
+            let trimmed_path_string = path_string.trim_start_matches("\\\\?\\");
+            if trimmed_path_string.is_empty() {
+                anyhow::bail!("File path is empty!");
```
Comment (SomeoneToIgnore) [inline crates/util/src/paths.rs:132]: I recall we were discussing file path length and other potential issues with such cut-offs, this is not what we're going to check and bail on here?

**Review by SomeoneToIgnore — COMMENTED** [review]
Interesting, I did not think of `PathWithPosition` for such case, maybe it's too much of a change but I'm not sure: check out how the error propagation works our as we have to fail on everything that tries to construct `PathWithPosition` now, propagating the error from `SanitizedPathBuf`'s potential failure during the constructions.

To me, it sounds a bit more coupling and ambitious to do, but not sure.
I think it's more realistic to adjust whatever workspace and other methods to require a sanitized path to cover more potential loopholes.

File: crates/util/src/paths.rs [hunk 5]
```diff
@@ -45,6 +46,9 @@ pub trait PathExt {
                 .ok_or_else(|| anyhow!("Invalid WTF-8 sequence: {bytes:?}"))
         }
     }
+
+    fn sanitized_pathbuf(&self) -> anyhow::Result<SanitizedPathBuf>;
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/util/src/paths.rs:50]: Thx!

File: crates/util/src/paths.rs [hunk 6]
```diff
@@ -93,6 +97,78 @@ impl<T: AsRef<Path>> PathExt for T {
...
+    }
+}
+
+impl Into<PathBuf> for SanitizedPathBuf {
+    fn into(self) -> PathBuf {
+        self.0
+    }
+}
+
+impl Into<SanitizedPathBuf> for PathBuf {
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/util/src/paths.rs:158]: In some places where I used `fn sanitized_pathbuf(&self) -> anyhow::Result<SanitizedPathBuf>`, handling the `Result` required additional steps, so I decided to wrap that handling into the `Into` implementation. However, upon reflection, I think there may be a better approach.

File: crates/util/src/paths.rs (hunk 1 above)
Comment (JunkuiZhang) [reply crates/util/src/paths.rs:142]: I'll give it another try and see what I can come up with, really thx!

File: crates/util/src/paths.rs [hunk 7]
```diff
@@ -45,6 +46,9 @@ pub trait PathExt {
                 .ok_or_else(|| anyhow!("Invalid WTF-8 sequence: {bytes:?}"))
         }
     }
+
+    fn sanitized_pathbuf(&self) -> anyhow::Result<SanitizedPathBuf>;
+    fn sanitized_pathbuf_string(&self) -> anyhow::Result<String>;
```
[the remark this answers is not in the document]
Comment (JunkuiZhang) [reply crates/util/src/paths.rs:51]: If I recall correctly, `fn sanitized_pathbuf_string(&self)` is used not only in `main.rs` but also in the Windows `platform.rs`. The reason for introducing this function was to reduce an extra layer of conversion from `PathBuf` to `String`.

However, considering that this function is not called frequently, it should be acceptable. I will go ahead and remove this function from `PathExt`.

File: crates/util/src/paths.rs (hunk 2 above)
Comment (JunkuiZhang) [reply crates/util/src/paths.rs:305]: Apologies for the outdated comment—I forgot to remove it.

File: crates/terminal_view/src/terminal_view.rs (hunk 3 above)
Comment (JunkuiZhang) [reply crates/terminal_view/src/terminal_view.rs:720]: > Overall, it's an interesting idea trying making `PathWithPosition`'s path sanitized.
> 
> Yet, original issue seemed to be in the fact that `workspace.open_paths` received canonicalized paths and failed to work with them properly? Any other "bad" input will still break things for workspace as the ctrl-click did before your fix. This is where we should sanitize too, and for that, we have to change `fn open_path` signature.

You’re right, the changes here are a simple “replacement” for #18726. `PathWithPosition` should retain `PathBuf`, and `SanitizedPathBuf` should be placed elsewhere.

> Maybe, same idea might relate to other changes in this PR? `trash_file`, `trash_dir`, and `file_save_dialog` seem all to accept a path, and can require a proper `SanitizedPathBuf` instead.
> 
> I find this direction interesting, as all of these require an absolute, symlink-traversed (i.e. canonicalized) path for their work, so `SanitizedPathBuf` looks natural there. More approchable and "natural" compared to placing the `SanitizedPathBuf` into `PathWithPosition`, as that one may accept relative paths and other things that look like paths but end with :22:33 things, so not really about the same `\\?\` issue?


My thinking is that for functions where we specifically want to use `SanitizedPathBuf` as a parameter, we should do so. However, for functions like `trash_file`, `trash_dir`, and `file_save_dialog`, we ideally want them to accept `PathBuf` with the `\\?\` prefix, but due to API limitations, we have to use `SanitizedPathBuf`. That’s why I didn’t change the function signatures.

File: crates/util/src/paths.rs (hunk 4 above)
Comment (JunkuiZhang) [reply crates/util/src/paths.rs:132]: My initial thought was that there might be cases like the following:

```rust
let path = PathBuf::from("\\\\?\\");
let result: SanitizedPathBuf = path.sanitized_path_buf();
```

However, I think it’s better to first check if the path is absolute, and only then perform the trim operation.

**Comment by CharlesChen0823** [comment]
IMO, using `PathWithPosition` should be limited for `terminal` and `cli`, don't using it for all crates. then we can control complex for deal with path. if user open project, there will not exists unc path?

**Review by SomeoneToIgnore — COMMENTED** [review]
Yes, this definitely gets out of hand and there seems to be no end to this.

In my original plan, I hoped we do not use the new type everywhere, but rather keep it on the border, to show very simplified as 
```
(bad paths are coming from here) ----------- being sanitized here ------------ again, regular path come out from workspace
------open_abs_path--------->|                                                |
------cmd_click------------->| workspace::foo(p: impl Into<SanitizedPathBuf>) | ----> calls below 
------smth_else------------->|                                                |
```

Seems that this would not work and we have to use the concrete type everywhere? Curious why cannot we keep the "regular" paths taken out of the sanitized ones after certain `workspace::foo` invocation.
Do you think it's worth pursuing this way?

**Comment by JunkuiZhang** [comment]
> Yes, this definitely gets out of hand and there seems to be no end to this.

Definitely.

> In my original plan, I hoped we do not use the new type everywhere, but rather keep it on the border, to show very simplified as
> 
> ```
> (bad paths are coming from here) ----------- being sanitized here ------------ again, regular path come out from workspace
> ------open_abs_path--------->|                                                |
> ------cmd_click------------->| workspace::foo(p: impl Into<SanitizedPathBuf>) | ----> calls below 
> ------smth_else------------->|                                                |
> ```
> 
> Seems that this would not work and we have to use the concrete type everywhere? Curious why cannot we keep the "regular" paths taken out of the sanitized ones after certain `workspace::foo` invocation.

I do want to keep the scope of changes as small as possible. Ultimately, I found that it seems all path-related issues can be addressed in one place: `WorktreeStore::find_or_create_worktree`. Of course, this PR has made more granular modifications to `WorktreeStore::find_worktree` and `WorktreeStore::create_worktree`.

As you can see, we ultimately use `some_abs_path` to create a `Worktree`. The question that arises is what kind of `PathBuf` should be stored in the `abs_path` of the `Worktree`—specifically, should it be a clean `PathBuf` without the `\\?\` prefix, or a "dirty" one?

The `Worktree::abs_path` is used in many places, but with different purposes: some calls to this function return a `PathBuf` intended for UI elements, like the tab view, which means the `abs_path` in the `Worktree` should hold a clean `PathBuf`. In contrast, other calls intend to use the `PathBuf` for I/O operations, so the `abs_path` should hold a dirty `PathBuf`.

Therefore, I think the `abs_path` in the `Worktree` should be of type `SanitizedPathBuf`, allowing the caller to decide whether to use a clean or dirty `PathBuf`. However, as you can see, this ultimately resulted in changes to over 1,000 lines of code.

> Do you think it's worth pursuing this way?

Honestly, I'm not quite sure. I need more time to experiment with different approaches and to reflect on the best solution. However, as I'm currently in the midterm period this semester, I have a lot of assignments and deadlines, so the progress on this PR may be slower than you expect. I will try to make some progress each day and communicate with you asynchronously.

**Pull request closed without merging**